### PR TITLE
DAOS-12398 object: Check punch epoch before migration punch (#11667)

### PR DESCRIPTION
**Pull request by wangdi1**
Since object enumeration might return next punch epoch of the current object/key/rec, so let's compare the punch epoch with rebuild/reintegration stable epoch before migrate punch.

Do not allow reint/drain/extend if there are other rebuild job going on to avoid screwing up the object layout.

Fix converting task error reset in dc_tx_check_update()

Retry UPDATE_AGAIN with delay, similar as INPROGRESS.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/control/dmg_pool_query_ranks.py [hunk 1]
```diff
@@ -134,8 +136,21 @@ def test_pool_query_ranks_mgmt(self):
...
+                    result = self.pool.reintegrate(rank)
+                except CommandFailure:
+                    self.log.debug("dmg command failed retry")
+                else:
+                    cmd_succeed = True
+                    break
+                counter += 1
+                time.sleep(3)
+
+            self.assertTrue(cmd_succeed, result)
```
Comment (daltonbohning) [inline src/tests/ftest/control/dmg_pool_query_ranks.py:153]: We really should supply a message when using asserts like this. Unfortunately, this is in master too
```suggestion
            self.assertTrue(cmd_succeed, result, "reason why the test failed")
```

**Review by daltonbohning — COMMENTED** [review]
We need to at least run the modified test:
```
Features: test_pool_query_ranks_error
```
and maybe some other feature tags?

**Review by liuxuezhao — DISMISSED** [review]

**Comment by jolivier23** [comment]
> We need to at least run the modified test:
> 
> ```
> Features: test_pool_query_ranks_error
> ```
> 
> and maybe some other feature tags?

Yeah, it appears that test didn't get run

**Comment by wangdi1** [comment]
> > We need to at least run the modified test:
> > ```
> > Features: test_pool_query_ranks_error
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and maybe some other feature tags?
> 
> Yeah, it appears that test didn't get run

sure, I will rerun

**Review by jolivier23 — COMMENTED** [review]
Needs merge approval.   I noticed Maureen commented on the ticket about an issue on master branch seen after the fix, I presume.   Is it related?

**Review by jolivier23 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Pull request merged**